src/json-api-koa.ts

```diff
@@ -0,0 +1,90 @@
+import { Context } from "koa";
+
+import Application from "./application";
+import { JsonApiDocument, Operation, OperationResponse } from "./types";
+
+export default function jsonApiKoa(app: Application) {
+  const URL_REGEX = new RegExp(app.types.map(t => t.name).join("|"), "i");
+
+  return async (ctx: Context, next: Function) => {
+    if (ctx.url.match(URL_REGEX) && app.types.length) {
+      const ops: Operation[] = convertHttpRequestToOperations(ctx);
+      const result: OperationResponse[] = await app.executeOperations(ops);
+      ctx.body = convertOperationsResponsesToHttpResponse(ctx, result);
+    } else if (ctx.url.match("/bulk")) {
+      return {
+        operations: await app.executeOperations(
+          ctx.request.body.operations || []
+        )
+      };
+    }
+
+    await next();
+  };
+}
+
+function convertHttpRequestToOperations(ctx: Context) {
+  const [, type, id] = ctx.url.split("/");
+
+  if (ctx.method === "GET") {
```

Comment:
@Exelord We can apply some DRY improvements to reduce this function a bit.

So, any HTTP method correlates to a JSONAPI operation:

```ts
const methodsToOperations = {
  GET: "get",
  POST: "add",
  DELETE: "remove",
  PATCH: "update",
  PUT: "update"
};
```

Every operation follows pretty much the same structure, except for `data`, which is present on the write operations, so we can do something like this:

```ts
const op = methodsToOperations[ctx.method];

return [
  {
    op,
    ref: { id, type },
    // Inject `data` only for write operations.
    ...(op === "create" || op === "update" ? { data: ctx.request.body.data } : {})
  } as Operation
];
```

Reply:
Oh yeah for sure! It's just the simplest way possible to make it works. I think we gonna start improving the implementation later, but it depends what Erik will say.